output/write.go

```diff
@@ -0,0 +1,112 @@
+package output
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+	"fmt"
+	"os"
+	"path/filepath"
+	"syscall"
+)
+
+// FileInfo returns the filesystem properties atomicWrite wrote
+type FileInfo struct {
+	Mode os.FileMode
+	UID  int
+	GID  int
+}
+
+// GetFileInfo from an open file
+func GetFileInfo(file *os.File) (*FileInfo, error) {
+	stat, err := file.Stat()
+	if err != nil {
+		return nil, fmt.Errorf("failed to stat after writing: %v", err)
+	}
+	filemode := stat.Mode()
+	uid := int(stat.Sys().(*syscall.Stat_t).Uid)
+	gid := int(stat.Sys().(*syscall.Stat_t).Gid)
+
+	return &FileInfo{filemode, uid, gid}, nil
+}
+
+// WriteFileAtomically creates a temporary file, sets perms, writes content, and renames it to filename
+// This sequence ensures the following:
+// 1. Nobody can open the file before we set owner/permissions properly
+// 2. Nobody observes a partially-overwritten secret file.
+// The returned FileInfo may not match the passed in one, especially if chownFiles is false.
+func WriteFileAtomically(dir, filename string, chownFiles bool, fileInfo FileInfo, enforceFilesystem Filesystem, content []byte) (*FileInfo, error) {
```

Comment:
This function takes a lot of arguments but they're all typed (except the first two) so I think it's okish